tests/unit/s2n_self_talk_state_machine_test.c

```diff
@@ -0,0 +1,123 @@
+
+        /* A connection cannot switch to the TLS12 state machine midway through the handshake */
+        {
+            EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+            EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+
+            /* Do handshake until the cert message is reached */
+            EXPECT_OK(s2n_negotiate_test_server_and_client_until_message(server_conn, client_conn, SERVER_CERT));
+
+            EXPECT_EQUAL(s2n_connection_get_actual_protocol_version(server_conn), S2N_TLS13);
+            EXPECT_EQUAL(s2n_connection_get_actual_protocol_version(client_conn), S2N_TLS13);
+
+            /* Alter which state machine the connection is using */
+            server_conn->actual_protocol_version = S2N_TLS12;
+            client_conn->actual_protocol_version = S2N_TLS12;
+
+            EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate_test_server_and_client(server_conn, client_conn), S2N_ERR_SAFETY);
+        }
+
+        /* A connection cannot switch to the TLS13 state machine midway through the handshake */
+        {
+            EXPECT_SUCCESS(s2n_connection_wipe(server_conn));
+            EXPECT_SUCCESS(s2n_connection_wipe(client_conn));
+            EXPECT_SUCCESS(s2n_connection_set_config(server_conn, tls12_config));
+            EXPECT_SUCCESS(s2n_connection_set_config(client_conn, tls12_config));
+            EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+            /* Do handshake until the server hello message is reached */
+            EXPECT_OK(s2n_negotiate_test_server_and_client_until_message(server_conn, client_conn, SERVER_HELLO));
+
+            EXPECT_EQUAL(s2n_connection_get_actual_protocol_version(server_conn), S2N_TLS12);
+            EXPECT_EQUAL(s2n_connection_get_actual_protocol_version(client_conn), S2N_TLS12);
+
+            /* Alter which state machine the connection is using */
+            server_conn->actual_protocol_version = S2N_TLS13;
+            client_conn->actual_protocol_version = S2N_TLS13;
+
+            EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate_test_server_and_client(server_conn, client_conn), S2N_ERR_SAFETY);
+        }
+
+        /* A hello retry handshake cannot change state machines */
+        {
+            EXPECT_SUCCESS(s2n_connection_wipe(server_conn));
+            EXPECT_SUCCESS(s2n_connection_wipe(client_conn));
+            EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+            EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+            EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+            /* Force the HRR path */
+            client_conn->security_policy_override = &security_policy_test_tls13_retry;
+
+            /* Negotiate handshake until cert message */
+            EXPECT_OK(s2n_negotiate_test_server_and_client_until_message(server_conn, client_conn, SERVER_CERT));
```

Comment:
Switching at the cert message skips what makes the HRR case interesting. We'd need to try to switch before the second ClientHello I think?

Maybe write the test so that we try to switch after every message?